pyhanko/sign/signers/csc_signer.py

```diff
@@ -0,0 +1,409 @@
+
+    def endpoint_url(self, endpoint_name):
+        return f"{self.service_url}/csc/{self.api_ver}/{endpoint_name}"
+
+    @property
+    def auth_headers(self):
+        return {'Authorization': f'Bearer {self.oauth_token}'}
+
+
+@dataclass(frozen=True)
+class CSCCredentialInfo:
+    signing_cert: x509.Certificate
+    chain: List[x509.Certificate]
+    supported_mechanisms: frozenset[str]
+    max_batch_size: int
+    hash_pinning_required: bool
+
+    def as_cert_store(self) -> CertificateStore:
+        scs = SimpleCertificateStore()
+        scs.register(self.signing_cert)
+        scs.register_multiple(self.chain)
+        return scs
+
+
+async def fetch_certs_in_csc_credential(session: aiohttp.ClientSession,
+                                        csc_session_info: CSCServiceSessionInfo,
+                                        timeout: int = 30) -> CSCCredentialInfo:
+    url = csc_session_info.endpoint_url("credentials/info")
+    req_data = {
+        "credentialID": csc_session_info.credential_id,
+        "certificates": "chain",
+        "certInfo": True
+    }
+
+    async with session.post(url=url, headers=csc_session_info.auth_headers,
+                            json=req_data, raise_for_status=True,
+                            timeout=timeout) as response:
+
+        try:
+            response_data = await response.json()
+        except aiohttp.ClientError as e:
+            raise SigningError("Credential info request failed") from e
+    return _process_certificate_info_response(response_data)
+
+
+def _process_certificate_info_response(response_data) -> CSCCredentialInfo:
+
+    try:
+        b64_certs = response_data['cert']['certificates']
+    except KeyError as e:
+        raise SigningError(
+            "Could not retrieve certificates from response"
+        ) from e
+    try:
+        certs = [
+            x509.Certificate.load(base64.b64decode(cert))
+            for cert in b64_certs
+        ]
+    except ValueError as e:
+        raise SigningError(
+            "Could not decode certificates in response"
+        ) from e
+    try:
+        algo_oids = response_data["key"]["algo"]
+        supported_algos = frozenset(
+            algos.SignedDigestAlgorithmId(oid).native for oid in algo_oids
+        )
+    except (KeyError, ValueError) as e:
+        raise SigningError(
+            "Could not retrieve supported signing mechanisms from response"
+        ) from e
+
+    try:
+        max_batch_size = int(response_data['multisign'])
+    except (KeyError, ValueError) as e:
+        raise SigningError(
+            "Could not retrieve max batch size from response"
+        ) from e
+
+    scal_value = int(response_data.get("SCAL", 1))
+    if scal_value not in (1, 2):
+        logger.warning(f"Unexpected SCAL value: {scal_value}; defaulting to 1")
+        scal_value = 1
+    hash_pinning_required = scal_value == 2
+
+    return CSCCredentialInfo(
+        signing_cert=certs[0], chain=certs[1:],
```

Comment:
A comment about cert zero being the signing cert is dropped. Not that it bothers me, but just pointing this out.

Reply:
Excellent point. I wrote this from scratch going off the spec, just trying to get to the point where I'd have something to play around with. I'll make sure to cross-check the parts that have functionality overlap with #29 as I add documentation & comments later.

Thanks! :)